### PR TITLE
Migrate from raw `hyper` server to `axum`

### PR DESCRIPTION
**Pull request by Turbo87**
This PR implements a `conduit-axum` compatibility layer, similar to `conduit-hyper`. This should allow us to incrementally migrate the existing middlewares and route handlers from `conduit` to [`axum`](https://github.com/tokio-rs/axum) in the future.

### CODE REVIEW

**Comment by Turbo87** [comment]
I've tried to reduce the diff as much as possible to ease the review process. The branch is currently deployed to staging for testing purposes and hasn't shown any issues so far.

I'd recommend review in [ignore whitespace mode](https://github.com/rust-lang/crates.io/pull/5605/files?w=1).

@jtgeibel @carols10cents @JohnTitor @pietroalbini @syphar if one of you has time to take a look at this I'd greatly appreciate it :)

**Comment by JohnTitor** [comment]
I'll review later today!

**Review by JohnTitor — APPROVED** [review]
The change and behavior on staging look great to me.

File: conduit-axum/src/fallback.rs [hunk 1]
```diff
@@ -19,68 +22,72 @@ use tracing::{error, warn};
...
+            .unwrap_or_else(|e| server_error_response(&e.to_string()))
+    })
+    .await
+    .map_err(Into::into)
 }
 
-/// Turns a `ConduitResponse` into a `HyperResponse`
-pub fn conduit_into_hyper(response: ConduitResponse) -> HyperResponse {
+/// Turns a `ConduitResponse` into a `AxumResponse`
+pub fn conduit_into_hyper(response: ConduitResponse) -> AxumResponse {
```
[the remark this answers is not in the document]
Comment (Turbo87) [reply conduit-axum/src/fallback.rs]: changed, thanks

File: conduit-axum/src/server.rs [hunk 2]
```diff
@@ -20,12 +15,8 @@ impl Server {
     /// control, such as configuring a graceful shutdown is necessary, then call
     /// `Service::from_blocking` instead.
     pub fn serve<H: conduit::Handler>(addr: &SocketAddr, handler: H) -> impl Future {
-        let handler = Arc::new(BlockingHandler::new(handler));
-        let make_service = make_service_fn(move |socket: &AddrStream| {
-            let handler = handler.clone();
-            let remote_addr = socket.remote_addr();
-            async move { Service::from_blocking(handler, remote_addr) }
-        });
+        let router = axum::Router::new().conduit_fallback(handler);
```
Comment (syphar) [inline conduit-axum/src/server.rs:18]: nit: 

later there will probably one place where you define the central routes, which means your `Router::new` should only be in one place, right?

Comment (syphar) [reply conduit-axum/src/server.rs:18]: Same for the middleware you apply, this also should be one place. So in tests & the normal server we have the same thing

**Review by syphar — COMMENTED** [review]
From what I know from axum from the docs.rs migration I don't see any problem here. 

I don't know the crates.io codebase much, so I might miss things that are specific to crates.io.

File: conduit-axum/src/server.rs (hunk 2 above)
Comment (Turbo87) [reply conduit-axum/src/server.rs:18]: yeah, you're right that with the `Server` struct we won't be able to adjust the axum router. The `Server` is only used by the `conduit-axum` examples and tests though, similar to how it was with `conduit-hyper` before. Our own code is using the slightly lower level APIs directly (see `build_handler()`), where we can add axum routes and middlewares without problems :)

**Pull request merged**